### PR TITLE
[feature] Add lines opening and closing times

### PR DESCRIPTION
**Pull request by ArnaudOggy**
Note : unit tests in another PR

### CODE REVIEW

File: src/model.rs [hunk 1]
```diff
@@ -606,6 +606,192 @@ impl Collections {
...
+        // Check if there is a need to calculate the opening/closing times.
+        // Generally it should be all or nothing (absent from a Gtfs, normally present if it's a recent Ntfs)
+        // In all cases a 2nd check is made below
+        let mut required_operation = false;
+        for line in self.lines.values() {
+            if line.opening_time.is_none() || line.closing_time.is_none() {
+                required_operation = true;
+                break;
+            }
+        }
```
Comment (woshilapin) [inline src/model.rs:744]:
```suggestion
        let required_operation = self                                               
            .lines                                                                  
            .values()                                                               
            .find(|line| line.opening_time.is_none() || line.closing_time.is_none())
            .is_some();                                                             
```

File: src/model.rs [hunk 2]
```diff
@@ -606,6 +606,192 @@ impl Collections {
...
+        if required_operation {
+            let vjs_by_line = get_vjs_by_line(&self);
+            let mut lines = self.lines.take();
+            for line in &mut lines {
+                // 2nd check (see above) to avoid overwriting line opening/closing
+                if line.opening_time.is_none() || line.closing_time.is_none() {
```
Comment (woshilapin) [inline src/model.rs:750]: You're doing twice the same test so maybe a closure can be extracted?

```rust
let check_time_empty = |line| line.opening_time.is_none() || line.closing_time.is_none();
```

File: src/model.rs [hunk 3]
```diff
@@ -606,6 +606,192 @@ impl Collections {
...
+                    let mut opening_timetable: TimeTable = BTreeMap::new();
+                    let mut closing_timetable: TimeTable = BTreeMap::new();
```
Comment (woshilapin) [inline src/model.rs:752]:
```suggestion
                    let mut opening_timetable = TimeTable::new();
                    let mut closing_timetable = TimeTable::new();
```

File: src/model.rs [hunk 4]
```diff
@@ -606,6 +606,192 @@ impl Collections {
...
+        // opening_timetable will be so {8: Time(29400), 9: Time(33000), 10: Time(36600)}
+        //
+        // For opening_timetable we keep the smallest schedule of the first departure for each time slot
+        // For closing_timetable we keep the biggest schedule of the last arrival for each time slot
+        fn fill_timetables(
+            vj: &VehicleJourney,
+            opening_timetable: &mut TimeTable,
+            closing_timetable: &mut TimeTable,
+        ) -> Result<()> {
+            let seconds_per_day: u32 = 86400;
```
[the remark this answers is not in the document]
Comment (ArnaudOggy) [reply src/model.rs]: Done in 2afc55e3d3f76d9b27d27691f8814870618710f7 (with constants)

File: src/model.rs [hunk 5]
```diff
@@ -606,6 +606,192 @@ impl Collections {
...
+
+        // Creates a map of (maximum) 24 elements, with possible indexes ranging from 0 to 23
+        // 2 timetables are created, one to store departures and the other for arrivals.
+        // Indeed, for a line/vehicle journey, the distance between two consecutive stops can take more than an hour.
+        //
+        // Example for a line with a vehicle journey leaving every half hour between 8:10am and 10:10am (6 vehicles journeys in total)
+        // opening_timetable will be so {8: Time(29400), 9: Time(33000), 10: Time(36600)}
+        //
+        // For opening_timetable we keep the smallest schedule of the first departure for each time slot
+        // For closing_timetable we keep the biggest schedule of the last arrival for each time slot
```
[the remark this answers is not in the document]
Comment (ArnaudOggy) [reply src/model.rs]: I revised the text. tell me if it's better. 
2afc55e3d3f76d9b27d27691f8814870618710f7

File: src/model.rs [hunk 6]
```diff
@@ -606,6 +606,192 @@ impl Collections {
...
+            opening_timetable: &mut TimeTable,
+            closing_timetable: &mut TimeTable,
+        ) -> Result<()> {
+            let seconds_per_day: u32 = 86400;
+            let vj_departure_time =
+                if let Some(vj_dpt_time) = vj.stop_times.first().map(|st| st.departure_time) {
+                    vj_dpt_time % seconds_per_day
+                } else {
+                    return Err(format_err!("undefined departure time for vj {}", vj.id));
+                };
```
[the remark this answers is not in the document]
Comment (ArnaudOggy) [reply src/model.rs]: Done in 2afc55e3d3f76d9b27d27691f8814870618710f7

File: src/model.rs [hunk 7]
```diff
@@ -606,6 +606,192 @@ impl Collections {
...
+
+        // Find the main hole for a line (i.e. without traffic), based on the timetable parameter.
+        // For example a line with vjs running from 04:10 to 22:45 will give this segment: [[23,24,0,1,2,3]]
+        // Several holes are possible in a day, such as: [[23,24,0,1,2,3],[12,13]]
+        // This function finds the largest and returns the two bounds/index (23 and 3 in the first example)
+        fn find_main_hole_boundaries(timetable: &TimeTable, hours_per_day: u8) -> Option<(u8, u8)> {
+            let mut holes: Vec<Vec<u8>> = Vec::new();
+            let mut is_last_elem_hole = false;
+            for i in 0..hours_per_day {
+                if !&timetable.contains_key(&i) {
```
[the remark this answers is not in the document]
Comment (ArnaudOggy) [reply src/model.rs]: Done in 2afc55e3d3f76d9b27d27691f8814870618710f7

File: src/model.rs (hunk 1 above)
Comment (ArnaudOggy) [reply src/model.rs]: Done in 2afc55e3d3f76d9b27d27691f8814870618710f7 (with any)

File: src/model.rs (hunk 2 above)
Comment (ArnaudOggy) [reply src/model.rs]: Done in 2afc55e3d3f76d9b27d27691f8814870618710f7

File: src/model.rs (hunk 3 above)
Comment (ArnaudOggy) [reply src/model.rs]: Done in 2afc55e3d3f76d9b27d27691f8814870618710f7

File: src/model.rs [hunk 8]
```diff
@@ -606,6 +606,192 @@ impl Collections {
...
+                    if !is_last_elem_hole {
+                        holes.push(vec![i]);
+                        is_last_elem_hole = true;
+                    } else if let Some(last) = holes.last_mut() {
+                        last.push(i)
+                    }
+                    // for the midnight passing, concatenate the vectors
+                    // [0,1] and [22,23] will become [22,23,0,1]
+                    if i == hours_per_day - 1 && holes.len() > 1 && holes[0].contains(&0) {
```
[the remark this answers is not in the document]
Comment (ArnaudOggy) [reply src/model.rs]: Not sure if it's panic. I was going to write the functional test. But it's better yes.
Done in cd0e341d5ff75bc898bf5773965817c57cc48f34

File: src/model.rs (hunk 1 above)
Comment (woshilapin) [reply src/model.rs]: Nice, indeed, `any` is more appropriate!

File: src/model.rs [hunk 9]
```diff
@@ -606,6 +606,192 @@ impl Collections {
...
+                    }
+                    None
+                })
+                .fold(HashMap::new(), |mut lines, (line_id, vj_idx)| {
+                    lines
+                        .entry(line_id)
+                        .or_insert_with(IdxSet::new)
+                        .insert(vj_idx);
+                    lines
+                })
```
[the remark this answers is not in the document]
Comment (ArnaudOggy) [reply src/model.rs:640]: Relevant but more general remark. We will discuss it together technically. Ticket created on the subject.

[the remark this answers is not in the document]
Comment (woshilapin) [reply src/model.rs:640]: Let's keep it like this for this PR and we'll improve if need be in a future PR.

File: src/documentation/common.md [hunk 10]
```diff
@@ -121,6 +121,12 @@ The following rules apply to every converter, unless otherwise explicitly specif
        `[name of origin's stop area] - [name of destination's stop area]`
   - the `route.destination_id` is set (overridden if needed) with the
     destination's stop area selected with the above rule
+- If a line has no opening and closing times, then they are generated.
+  - if one of the two values (at least) is empty, then both are recalculated, otherwise they are kept.
```
Comment (woshilapin) [inline src/documentation/common.md:125]:
```suggestion
- If a line has an empty opening or closing times, then they are both generated.
```

Comment (ArnaudOggy) [reply src/documentation/common.md]: Commited :+1:

**Pull request merged**